hedera-node/hedera-app/src/main/java/com/hedera/node/app/tss/handlers/TssVoteHandler.java

```diff
@@ -53,5 +62,71 @@ public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckExcept
+        if (tssBaseStore.exists(tssVoteMapKey)) {
+            // Duplicate vote
+            return;
+        }
+
+        if (!TssVoteHandler.hasReachedThreshold(txBody, context)) {
+            tssBaseStore.put(tssVoteMapKey, txBody);
+        }
+    }
+
+    /**
+     * Check if the threshold number of votes (totaling at least 1/3 of weight) have already been received for the
+     * candidate roster, all with the same vote byte array.
+     *
+     * @param tssVoteTransaction the TssVoteTransaction to check
+     * @param context the HandleContext
+     * @return true if the threshold has been reached, false otherwise
+     */
+    public static boolean hasReachedThreshold(
+            @NonNull final TssVoteTransactionBody tssVoteTransaction, @NonNull final HandleContext context) {
+        final var rosterStore = context.storeFactory().readableStore(ReadableRosterStore.class);
+
+        final Roster activeRoster = rosterStore.getActiveRoster();
+        if (activeRoster == null) {
+            throw new IllegalArgumentException("No active roster found");
+        }
+
+        // Get all votes for the active roster
+        final Map<RosterEntry, TssVoteTransactionBody> voteByNode = new HashMap<>();
+
+        // Get the target roster from the TssVoteTransactionBody
+        final Bytes targetRosterHash = tssVoteTransaction.targetRosterHash();
+
+        // Also get the total active roster weight
+        long activeRosterTotalWeight = 0;
+
+        final var tssBaseStore = context.storeFactory().writableStore(WritableTssStore.class);
+        // For every node in the active roster, check if there is a vote for the target roster hash
+        for (final RosterEntry rosterEntry : rosterStore.getActiveRoster().rosterEntries()) {
+            activeRosterTotalWeight += rosterEntry.weight();
+            final TssVoteMapKey tssVoteMapKey = new TssVoteMapKey(targetRosterHash, rosterEntry.nodeId());
+            if (tssBaseStore.exists(tssVoteMapKey)) {
+                voteByNode.put(rosterEntry, tssBaseStore.getVote(tssVoteMapKey));
+            }
+        }
+
+        // Initialize a counter for the total weight of votes with the same vote byte array
+        long voteWeight = 0L;
+
+        // Iterate over the votes which has the same target roster hash
+        for (final RosterEntry rosterEntryKey : voteByNode.keySet()) {
+            final TssVoteTransactionBody vote = voteByNode.get(rosterEntryKey);
+            // If the vote byte array matches the one in the TssVoteTransaction, add the weight of the vote to the
+            // counter
+            if (vote.tssVote().equals(tssVoteTransaction.tssVote())) {
+                voteWeight += rosterEntryKey.weight();
+            }
+        }
+
+        // Check if the total weight of votes with the same vote byte array is at least 1/3 of the
+        // total weight of the network
+        return voteWeight >= activeRosterTotalWeight / 3;
```

Comment:
Integer division rounds down.  There is a corner case here where less than 1/3 vote by weight is allowing the candidate roster to become adopted.  Add 1 to the value to avoid the corner case.  The most accurate solution would add 1 when the modulus is non-zero.

Reply:
Agreed. Fixed. Thanks @edward-swirldslabs